addons/mail/models/mail_thread_main_attachment.py

```diff
@@ -0,0 +1,37 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class MailMainAttachmentMixin(models.AbstractModel):
+    """ Mixin that is inherited by all model with main attachment.
+    """
+
+    _name = 'mail.thread.main.attachment'
+    _inherit = ['mail.thread']
+    _description = 'Mail Main Attachment Management Mixin'
+
+    message_main_attachment_id = fields.Many2one(string="Main Attachment", comodel_name='ir.attachment', copy=False)
+
+    def _message_post_after_hook(self, message, msg_values):
+        """ Set main attachment field if necessary """
+        super()._message_post_after_hook(message, msg_values)
+        self._message_set_main_attachment_id([
+            attachment_command[1]
+            for attachment_command in (msg_values['attachment_ids'] or [])
+        ])
+
+    def _message_set_main_attachment_id(self, attachment_ids):
+        if attachment_ids and not self.message_main_attachment_id:
+            all_attachments = self.env['ir.attachment'].browse(attachment_ids)
+            prioritary_attachments = all_attachments.filtered(lambda x: x.mimetype.endswith('pdf')) \
+                                     or all_attachments.filtered(lambda x: x.mimetype.startswith('image')) \
+                                     or all_attachments
+            self.sudo().with_context(tracking_disable=True).message_main_attachment_id = prioritary_attachments[0].id
```

Comment:
While we are at it, this ``sudo`` is annoying me since I saw it ... but indeed it seems you can have the right to post on a document without having the write access (e.g. mail_post_access set to read, or replying to a notification on a document you can't read). It comes from https://github.com/odoo/odoo/pull/30659 .

I have a challenge (entrepreneur version) or a request (human version) for you: could you write a small unit test for this sudo ? I think we already have either a record with mail_post_access tweak, either an example of patched ACL in some mail tests (sorry for not being more precise, have to dig into those files :D ) . That would be great :)

Reply:
As discussed, it has been moved to a child task (3178885) for not stalling this one, see odoo/odoo#112273.